File: example_1.m
% Example script to find locations and times of minimum and maximum values

clear all, close all

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% load the data


file_name='sst.mnmean.nc';

sst=ncread(file_name,'sst');
lon_sst=ncread(file_name,'lon');
lat_sst=ncread(file_name,'lat');
time_sst=ncread(file_name,'time');

time_sst=time_sst+datenum(1800,1,1);
time2_sst=datestr(time_sst,'mmm yyyy');

mask=ncread('lsmask_sst.nc','mask');
mask3=repmat(mask,[1 1 size(sst,3)]);
sst(mask3==0)=NaN; clear mask3

sst=permute(sst,[3 2 1]);


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Find the maximum. You can easily modify the code below to find the
% location and time of the minimum...


% Average over three months:
sst_rm=runmean_maps(sst,3);

% Identify the maximum at each location
[max_sst] = squeeze(max(sst_rm,[],1));

% Find the maximum over all longitudes as a function of latitude:
[max_sst_lat] = max(max_sst,[],2);

% Get the latitude of the maximum temperature:
[~,ilat]=max(max_sst_lat); max_lat=lat_sst(ilat);

% Then you find the maximum longitude with:
max_sst_lon=squeeze(max_sst(ilat,:));
[~,ilon]=max(max_sst_lon); max_lon=lon_sst(ilon);

% ... and the time:
max_sst_time=squeeze(sst_rm(:,ilat,ilon));
[~,itime]=max(max_sst_time); max_time=time_sst(itime);

% check if you have really found the right location and time:
check=isequal(squeeze(max(max(max(sst_rm)))),...
    squeeze(sst_rm(itime,ilat,ilon)));

if check==1
    disp(['The maximum occurred at Latitude = ' num2str(max_lat) ' ' ...
        char(176) 'N and Longitude = ' ...
        num2str(max_lon) ' ' char(176) 'E around ' time2_sst(itime,:) '.'])
elseif check==0
    disp('Something went wrong.')
end


% See what country or ocean is at that location:


figure(1)
[~,h]=contourf(lon_sst,lat_sst,max_sst,50);
set(h,'EdgeColor','none')
colorbar
caxis([0 30])
set(gca,'FontSize',14)
hold on
contourf(lon_sst,lat_sst,-squeeze(mask)',[0 0],'facecolor',[0.7 0.7 0.7],'color','k','linewidth',2)
%contour(lon_sst,lat_sst,-squeeze(mask)',[0 0],'color','k','linewidth',2)
plot(max_lon,max_lat,'rx','linewidth',4,'markersize',12)
hold off
xlabel('Longitude (\circE)')
ylabel('Latitude (\circN)')
title('Maximum SST (\circC)')


% If you would like to add a coastline to the air temperature and
% precipitation plots, you can use the 'land.nc' data. If you need help,
% just ask me.

File: Python_Scripts/example_1.py
# Example script to find locations and times of minimum and maximum values

#Import
from pathlib import Path
import os
import xarray as xr
import matplotlib.pyplot as plt

#%% Import the data
data_folder = Path(os.getcwd()) / 'Data'
file_name='sst.mnmean.nc'
sst = xr.open_dataset(data_folder / file_name).sst
mask = xr.open_dataarray(data_folder / 'lsmask_sst.nc')
sst_masked = sst.where(mask.values == 1)

#now apply a running mean filter to smooth the dataset
sst_rm = sst_masked.rolling(time=3).mean()
# Identify the maximum at each location over time
max_sst = sst_rm.max(dim = 'time')
# Find the maximum over all longitudes as a function of latitude:
max_sst_lat = max_sst.max(dim = 'lon')
# Get the latitude of the maximum temperature:
max_lat = max_sst_lat.idxmax('lat').values.item() # see https://stackoverflow.com/questions/40179593/how-to-get-the-coordinates-of-the-maximum-in-xarray
# note that .item converts 1D ndarray into scalar, see here: https://numpy.org/doc/stable/reference/generated/numpy.ndarray.item.html

# now find the maximum longitude
max_lon = max_sst.max(dim = 'lat').idxmax('lon').values.item()

#now find the maximum over time at the identified location
#TODO
max_sst_time = sst_rm[:,max_lat,max_lon].max(dim='time').values.item()

# See what country or ocean is at that location:

#%% Some random plots
# Maximum SST as a function of latitude
fig1, ax1 = plt.subplots(nrows=2)
fig1.subplots_adjust(hspace=.3)
max_sst_lat.plot(ax=ax1[0])
max_sst.max(dim = 'lat').plot(ax=ax1[1])
fig1.show()

